src/details/second_order_lowpass.cpp

```diff
@@ -1,36 +1,48 @@
 #include <cmath>
+#include <limits>
 #include <path_tracking_pid/details/second_order_lowpass.hpp>
 
 namespace path_tracking_pid::details
 {
-namespace
-{
-// Used in filter calculations. Default 1.0 corresponds to a cutoff frequency at 1/4 of the sample rate.
-constexpr double cutoff = 1.;
+constexpr auto NaN = std::numeric_limits<float>::quiet_NaN();
 
-}  // namespace
+SecondOrderLowpass::SecondOrderLowpass() : cutoff_(NaN), damping_(NaN) {}
 
-void SecondOrderLowpass::push(double value)
+SecondOrderLowpass::SecondOrderLowpass(double cutoff, double damping)
+: cutoff_(cutoff), damping_(damping)
 {
-  errors_.push(value);
-
-  filtered_errors_.push(
-    (1 / (1 + cutoff * cutoff + M_SQRT2 * cutoff)) *
-    (errors_.at<2>() + 2 * errors_.at<1>() + errors_.at<0>() -
-     (cutoff * cutoff - M_SQRT2 * cutoff + 1) * filtered_errors_.at<1>() -
-     (-2 * cutoff * cutoff + 2) * filtered_errors_.at<0>()));
 }
 
-void SecondOrderLowpass::reset()
+void SecondOrderLowpass::configure(double cutoff, double damping)
 {
-  errors_.reset();
-  filtered_errors_.reset();
+  cutoff_ = cutoff;
+  damping_ = damping;
 }
 
-const FifoArray<double, 3> & SecondOrderLowpass::errors() const { return errors_; }
-const FifoArray<double, 3> & SecondOrderLowpass::filtered_errors() const
+double SecondOrderLowpass::filter(double u, double step_size)
 {
-  return filtered_errors_;
+  // save history
+  u_.push(u);
+  y_.push(u);  // increase index so the math below looks correct
+
+  if (cutoff_ == 0) {
+    return u;
+  }
+
+  auto c = cutoff_;
```

Comment:
Consider making constants actually `const`. That should be your default. Only remove it, when required.

Also see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-immutable

Reply:
Hmm I don't know. For me adding const here adds clutter. It doesn't add that much because the scope is so small. I can understand from a "purity" type of view this is wrong, but for readability it is good.

Using const in API's I'm fully onboard